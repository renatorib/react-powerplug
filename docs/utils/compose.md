# compose

The compose utility is a factory version of [Compose](/docs/components/Compose.md) component.

```js
import { compose, Counter, Toggle } from 'react-powerplug' // note lowercased (c)ompose

// the order matters
const ToggleCounter = compose(
  <Counter initial={5} />, // accept a element
  Toggle,                  // or just a component
)

<ToggleCounter>
  {(counter, toggle) => {
    // counter.inc, counter.dec, counter.count
    // toggle.on, toggle.toggle, etc.
  }}
</ToggleCounter>
```

---

Behind the scenes, that's what happens:

```jsx
<Counter initial={5}>
  {counter => (
    <Toggle>
      {toggle => (
        /* ... */
      )}
    </Toggle>
  )}
</Counter>
```
