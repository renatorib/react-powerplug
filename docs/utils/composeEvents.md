# composeEvents

The composeEvents utility helps you when you need to pass the same callback more than once.

```jsx
import { Hover, composeEvents } from 'react-powerplug'

const HoveredDiv = ({ children, onMouseEnter, onMouseLeave, ...restProps }) => (
  <Hover>
    {({ isHover, bindHover }) => (
      <div
        {...composeEvents({ onMouseEnter, onMouseLeave }, bindHover)}
        {...restProps}
        children={children(isHover)}
      />
    )}
  </Hover>
)
```

It's just merge array of events object into single one.

```jsx
const callbacks = composeEvents(
  {
    onMouseEnter: event => console.log('first call', event),
    onMouseLeave: event => console.log('first call', event),
  },
  {
    onMouseEnter: event => console.log('second call', event),
  }
)

/**
 * callbacks = {
 *   onMouseEnter: Function,
 *   onMouseLeave: Function
 * }
 */

const App = () => (
  <>
    <div {...callbacks} />
    <div onMouseEnter={callbacks.onMouseEnter} />
  </>
)
```
