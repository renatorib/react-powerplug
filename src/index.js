export { default as Active } from './components/Active'
export { default as Compose } from './components/Compose'
export { default as Counter } from './components/Counter'
export { default as Field } from './components/Field'
export { default as Focus } from './components/Focus'
export { default as unstable_FocusManager } from './components/FocusManager'
export { default as Form } from './components/Form'
export { default as Hover } from './components/Hover'
export { default as Interval } from './components/Interval'
export { default as List } from './components/List'
export { default as Map } from './components/Map'
export { default as Set } from './components/Set'
export { default as State } from './components/State'
export { default as Toggle } from './components/Toggle'
export { default as Touch } from './components/Touch'
export { default as Value } from './components/Value'

export { default as compose } from './utils/compose'
export { default as composeEvents } from './utils/composeEvents'
export { default as renderProps } from './utils/renderProps'
