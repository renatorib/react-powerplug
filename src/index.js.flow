// @flow

import * as React from 'react'

/* Active */

type ActiveChange = ({| isActive: boolean |}) => void

type ActiveRender = ({|
  isActive: boolean,
  bindActive: {| onMouseDown: () => void, onMouseUp: () => void |},
|}) => React.Node

declare export var Active: React.ComponentType<
  | {| onChange?: ActiveChange, render: ActiveRender |}
  | {| onChange?: ActiveChange, children: ActiveRender |}
>

/* Bind */

type BindChange = ({| value: string |}) => void

type BindRender = ({|
  value: string,
  setValue: string => void,
  bind: {| value: string, onChange: (SyntheticInputEvent<*>) => void |},
|}) => React.Node

declare export var Bind: React.ComponentType<
  | {| initial?: string, onChange?: BindChange, render: BindRender |}
  | {| initial?: string, onChange?: BindChange, children: BindRender |}
>

/* Compose */

declare export var Compose: React.ComponentType<Object>

/* Counter */

type CounterChange = ({| count: number |}) => void

type CounterRender = ({|
  count: number,
  inc: (step?: number) => void,
  dec: (step?: number) => void,
|}) => React.Node

declare export var Counter: React.ComponentType<
  | {| initial?: number, onChange?: CounterChange, render: CounterRender |}
  | {| initial?: number, onChange?: CounterChange, children: CounterRender |}
>

/* Focus */

type FocusChange = ({| isFocus: boolean |}) => void

type FocusRender = ({|
  isFocus: boolean,
  bindFocus: {| onFocusIn: () => void, onFocusOut: () => void |},
|}) => React.Node

declare export var Focus: React.ComponentType<
  | {| onChange?: FocusChange, render: FocusRender |}
  | {| onChange?: FocusChange, children: FocusRender |}
>

/* Form */

type FormChange<T> = T => void

type FormRender<T> = ({|
  input: <K: $Keys<T>>(
    key: K
  ) => {|
    value: string,
    setValue: string => void,
    bind: {|
      value: string,
      onChange: (SyntheticInputEvent<*>) => void,
    |},
  |},
|}) => React.Node

type FormProps<T> =
  | {| initial: T, onChange?: FormChange<T>, render: FormRender<T> |}
  | {| initial: T, onChange?: FormChange<T>, children: FormRender<T> |}

declare export class Form<T: { [string]: string }> extends React.Component<
  FormProps<T>,
  *
> {}

/* Hover */

type HoverChange = ({| isHover: boolean |}) => void

type HoverRender = ({|
  isHover: boolean,
  bindHover: {| onMouseEnter: () => void, onMouseLeave: () => void |},
|}) => React.Node

declare export var Hover: React.ComponentType<
  | {| onChange?: HoverChange, render: HoverRender |}
  | {| onChange?: HoverChange, children: HoverRender |}
>

/* List */

type ListChange<T> = ({| list: $ReadOnlyArray<T> |}) => void

type ListRender<T> = ({|
  list: $ReadOnlyArray<T>,
  setList: ($ReadOnlyArray<T>) => void,
  push: (value: T) => void,
  pull: (predicate: (T) => boolean) => void,
  sort: (compare: (a: T, b: T) => -1 | 0 | 1) => void,
|}) => React.Node

type ListProps<T> =
  | {|
      initial: $ReadOnlyArray<T>,
      onChange?: ListChange<T>,
      render: ListRender<T>,
    |}
  | {|
      initial: $ReadOnlyArray<T>,
      onChange?: ListChange<T>,
      children: ListRender<T>,
    |}

declare export class List<T> extends React.Component<ListProps<T>, *> {}

/* Loading */

type LoadingChange = ({| isLoading: boolean |}) => void

type LoadingRender = ({|
  isLoading: boolean,
  toggleLoading: () => void,
  setLoading: (isLoading: boolean) => void,
|}) => React.Node

declare export var Loading: React.ComponentType<
  | {| initial?: boolean, onChange?: LoadingChange, render: LoadingRender |}
  | {| initial?: boolean, onChange?: LoadingChange, children: LoadingRender |}
>

/* Set */

type SetChange<T> = ({| values: $ReadOnlyArray<T> |}) => void

type SetRender<T> = ({|
  values: $ReadOnlyArray<T>,
  add: (key: T) => void,
  clear: () => void,
  remove: (key: T) => void,
  has: (key: T) => boolean,
|}) => React.Node

type SetProps<T> =
  | {|
      initial: $ReadOnlyArray<T>,
      onChange?: SetChange<T>,
      render: SetRender<T>,
    |}
  | {|
      initial: $ReadOnlyArray<T>,
      onChange?: SetChange<T>,
      children: SetRender<T>,
    |}

declare export class Set<T> extends React.Component<SetProps<T>> {}

/* Map */

type MapChange<T> = T => void

type MapRender<T, K = $Keys<T>> = ({|
  values: T,
  set: (key: K, value: $ElementType<T, K>) => void,
  over: (key: K, fn: ($ElementType<T, K>) => $ElementType<T, K>) => void,
  get: (key: K) => $ElementType<T, K>,
|}) => React.Node

type MapProps<T> =
  | {| initial: T, onChange?: MapChange<T>, render: MapRender<T> |}
  | {| initial: T, onChange?: MapChange<T>, children: MapRender<T> |}

declare export class Map<T: Object> extends React.Component<MapProps<T>, *> {}

/* State */

type StateChange<T> = T => void

type StateRender<T> = ({|
  state: T,
  setState: (updated: $Shape<T> | (T => $Shape<T>), cb?: () => void) => void,
|}) => React.Node

type StateProps<T> =
  | {| initial: T, onChange?: StateChange<T>, render: StateRender<T> |}
  | {| initial: T, onChange?: StateChange<T>, children: StateRender<T> |}

declare export class State<T: Object> extends React.Component<StateProps<T>> {}

/* Toggle */

type ToggleChange = ({| on: boolean |}) => void

type ToggleRender = ({|
  on: boolean,
  off: boolean,
  toggle: () => void,
  setOn: (on: boolean) => void,
|}) => React.Node

declare export var Toggle: React.ComponentType<
  | {| initial?: boolean, onChange?: ToggleChange, render: ToggleRender |}
  | {| initial?: boolean, onChange?: ToggleChange, children: ToggleRender |}
>

/* Value */

type ValueChange<T> = ({| value: T |}) => void

type ValueRender<T> = ({|
  value: T,
  setValue: T => void,
|}) => React.Node

type ValueProps<T> =
  | {| initial: T, onChange?: ValueChange<T>, render: ValueRender<T> |}
  | {| initial: T, onChange?: ValueChange<T>, children: ValueRender<T> |}

declare export class Value<T> extends React.Component<ValueProps<T>> {}
