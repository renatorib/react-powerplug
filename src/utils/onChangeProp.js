import noop from './noop'

const onChangeProp = (originalOnChange = noop, propName) => state => {
  originalOnChange(state[propName])
}

export default onChangeProp
