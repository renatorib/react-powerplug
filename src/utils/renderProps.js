import warn from './warn'

const isFn = prop => typeof prop === 'function'

/**
 * renderProps
 * is a render/children props interop.
 * will pick up the prop that was used,
 * or children if both are used
 */

const renderProps = ({ children, render }, ...props) => {
  if (process.env.NODE_ENV !== 'production') {
    warn(
      isFn(children) && isFn(render),
      'You are using the children and render props together.\n' +
        'This is impossible, therefore, only the children will be used.'
    )
  }

  const fn = isFn(children) ? children : render

  return fn ? fn(...props) : null
}

export default renderProps
