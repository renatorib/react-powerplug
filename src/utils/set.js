export const set = (updater, arg) =>
  typeof updater === 'function' ? updater(arg) : updater
