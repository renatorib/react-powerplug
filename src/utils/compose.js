import * as React from 'react'
import renderProps from './renderProps'

const isElement = element => typeof element.type === 'function'

const compose = (...elements) => {
  const reversedElements = elements.reverse()

  return composedProps => {
    // Stack children arguments recursively and pass
    // it down until the last component that render children
    // with these stacked arguments
    function stackProps(i, elements, propsList = []) {
      const element = elements[i]
      const isTheLast = i === 0

      // Check if is latest component.
      // If is latest then render children,
      // Otherwise continue stacking arguments
      const renderFn = props =>
        isTheLast
          ? renderProps(composedProps, ...propsList.concat(props))
          : stackProps(i - 1, elements, propsList.concat(props))

      // Clone a element if it's passed created as <Element initial={} />
      // Or create it if passed as just Element
      const elementFn = isElement(element)
        ? React.cloneElement
        : React.createElement

      return elementFn(element, {}, renderFn)
    }

    return stackProps(elements.length - 1, reversedElements)
  }
}

export default compose
