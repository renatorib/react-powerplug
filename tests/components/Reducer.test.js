import * as React from 'react'
import TestRenderer from 'react-test-renderer'
import { Reducer } from '../../src'
import { last } from './utils'

test('should handle dispatched actions', () => {
  const renderFn = jest.fn().mockReturnValue(null)

  const initial = { count: 0, somethingElse: 'whatever' }

  const actions = {
    incrementCount: (state, step = 1) => ({ count: state.count + step }),
  }

  TestRenderer.create(
    <Reducer actions={actions} initial={initial} render={renderFn} />
  )

  // Initial values
  expect(renderFn).lastCalledWith({
    state: { ...initial },
    dispatch: expect.any(Function),
  })

  last(renderFn.mock.calls)[0].dispatch('incrementCount', 10)
  expect(renderFn).lastCalledWith({
    state: { count: 10, somethingElse: 'whatever' },
    dispatch: expect.any(Function),
  })
})

test('should error if no action matches dispatched type', () => {
  const renderFn = jest.fn().mockReturnValue(null)

  TestRenderer.create(<Reducer render={renderFn} />)

  expect(() =>
    last(renderFn.mock.calls)[0].dispatch('notAnAction')
  ).toThrowErrorMatchingSnapshot()
})
