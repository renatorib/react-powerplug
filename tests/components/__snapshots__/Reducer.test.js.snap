// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should error if no action matches dispatched type 1`] = `"No handler is defined for this type of action"`;
